transformations/transformations/data_offload.py

```diff
@@ -613,3 +614,274 @@ def process_driver(self, routine, successors):
+        # add imports for symbols hoisted
+        symbol_map = defaultdict(set)
+        for var, module in chain(all_uses_symbols, all_defines_symbols):
+            # filter modules that are supposed to be ignored
+            if module.lower() in self.ignore_modules:
+                continue
+            symbol_map[module].add(var.parents[0] if var.parent else var)
+        import_map = CaseInsensitiveDict()
+        scope = routine
+        while scope:
+            import_map.update(scope.import_map)
+            scope = scope.parent
+        missing_imports_map = defaultdict(set)
+        for module, variables in symbol_map.items():
+            missing_imports_map[module] |= {var for var in variables if var.name not in import_map}
+        if missing_imports_map:
+            routine.spec.prepend(Comment(text=(
+                '![Loki::GlobalVarHoistTransformation] ---------------------------------------'
+            )))
+            for module, variables in missing_imports_map.items():
+                symbols = tuple(var.clone(dimensions=None, scope=routine) for var in variables)
+                routine.spec.prepend(Import(module=module, symbols=symbols))
+
+            routine.spec.prepend(Comment(text=(
+                '![Loki::GlobalVarHoistTransformation] '
+                '-------- Added global variable imports for offload directives -----------'
+            )))
+
+    def process_kernel(self, routine, successors, item):
+        """
+        Hoist mdule variables for kernel routines.
+
+        This includes: appending the corresponding variables
+        to the routine arguments as well as to calls within the kernel
+        and removing the imports that became unused.
+        """
+        # get symbols per routine (successors)
+        defines_symbols, uses_symbols = self._get_symbols(successors)
+
+        # append symbols to routine (arguments)
+        self._append_routine_arguments(routine, item)
+
+        # append symbols to calls (arguments)
+        self._append_call_arguments(routine, uses_symbols, defines_symbols)
+
+        # get symbols for this routine/kernel
+        kernel_defines_symbols = item.trafo_data.get(self._key, {}).get('defines_symbols', set())
+        kernel_uses_symbols = item.trafo_data.get(self._key, {}).get('uses_symbols', set())
+        # remove imports for symbols hoisted
+        symbol_map = defaultdict(set)
+        for var, module in chain(kernel_uses_symbols, kernel_defines_symbols):
+            # filter modules that are supposed to be ignored
+            if module.lower() in self.ignore_modules:
+                continue
+            symbol_map[module].add(var.parents[0] if var.parent else var)
+        import_map = CaseInsensitiveDict(
+                (s.name, imprt) for imprt in routine.all_imports[::-1] for s in imprt.symbols
+                )
+        scope = routine
+        while scope:
+            import_map.update(scope.import_map)
+            scope = scope.parent
```

Comment:
Sorry for being pedantic, but this bit needs to go now otherwise you're still overwriting the values with parent scope entries ;-)
```suggestion
```

Reply:
... yes ... fair enough 😅